File: 백준/Gold/12908. 텔레포트 3/텔레포트 3.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		pairs = new Pair[8];
		st = new StringTokenizer(br.readLine());
		sx = Integer.parseInt(st.nextToken());
		sy = Integer.parseInt(st.nextToken());
		pairs[idx] = new Pair(sx,sy);
		idx++;
		
		st = new StringTokenizer(br.readLine());
		ex = Integer.parseInt(st.nextToken());
		ey = Integer.parseInt(st.nextToken());
		pairs[idx] = new Pair(ex,ey);
		idx++;
		dist = new long[8][8];
		for(long row[] : dist) {
			Arrays.fill(row, INF);
		}
		for(int a = 0; a < 3; a++) {
			st = new StringTokenizer(br.readLine());
			int x1 = Integer.parseInt(st.nextToken());
			int y1 = Integer.parseInt(st.nextToken());
			pairs[idx] = new Pair(x1,y1);
			int i = idx;
			idx += 1;
			int x2 = Integer.parseInt(st.nextToken());
			int y2 = Integer.parseInt(st.nextToken());
			pairs[idx] = new Pair(x2,y2);
			int j = idx;
			idx += 1;
			
			dist[i][j] = 10;
			dist[j][i] = 10;
		}
		
		for(int i = 0; i < 8; i++) {
			for(int j = 0; j < 8; j++) {
				if(i == j) {
					dist[i][j] = dist[j][i] = 0;
					continue;
				}
				
				Pair pair1 = pairs[i];
				Pair pair2 = pairs[j];
				
				long distance = Math.abs(pair1.x - pair2.x) + Math.abs(pair1.y - pair2.y);
				dist[i][j] = Math.min(dist[i][j], distance);
				dist[j][i] = Math.min(dist[j][i], distance);
			}
		}
		
		for(int k = 0; k < 8; k++) {
			for(int i = 0; i < 8; i++) {
				for(int j = 0; j< 8; j++) {
					dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
				}
			}
		}
		
		System.out.println(dist[0][1]);
		
	}
	static long dist[][];
	static long INF = Long.MAX_VALUE / 2;
	static StringTokenizer st;
	static int idx;
	static Pair pairs[];
	static int sx,sy,ex,ey;
	
	static class Pair{
		int x,y;
		public Pair(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}
}

File: 백준/Gold/12908. 텔레포트 3/텔레포트 3.py
MAX_R = 1_000_000_000 * 4
def printBoard(board, width=5):

    for y in range(8):
        for x in range(8):
            if board[y][x] == MAX_R:
                print(f"{'INF':>{width}}", end='')
            else:
                print(f"{board[y][x]:>{width}}",end='')
        print()
if __name__ == "__main__":

    idx2pair = dict()
    idx = 0

    sx, sy = map(int,input().split())

    idx2pair[idx] = (sx,sy)
    idx += 1
    ex, ey = map(int,input().split())
    idx2pair[idx] = (ex,ey)
    idx += 1

    dist = [[MAX_R for _ in range(8)] for _ in range(8)]

    for _ in range(3):
        x1, y1, x2, y2 = map(int,input().split())
        idx2pair[idx] = (x1,y1)
        i = idx
        idx += 1

        j = idx
        idx2pair[idx] = (x2,y2)
        idx += 1

        dist[i][j] = 10
        dist[j][i] = 10


    # printBoard(dist)
    # print('----')
    for i in range(8):
        for j in range(8):
            if i == j:
                dist[i][j] = dist[j][i] = 0
                continue
            x1,y1 = idx2pair[i]
            x2,y2 = idx2pair[j]
            distance = abs(x1-x2) + abs(y1-y2)

            dist[i][j] = min(dist[i][j], distance)
            dist[j][i] = min(dist[j][i], distance)

    # printBoard(dist)
    # print('----')
    #
    for k in range(8):
        for i in range(8):
            for j in range(8):
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

    # printBoard(dist)

    print(dist[0][1])







